src/command.ts

```diff
@@ -0,0 +1,261 @@
+  private helpText = "";
+  private client: any;
+
+  /**
+   * @param cmd the command to create.
+   */
+  constructor(private cmd: string) {
+    this.name = first(cmd.split(" ")) || "";
+    this.options = [];
+    this.befores = [];
+  }
+
+  /**
+   * Sets the description of the command.
+   * @param t a human readable description.
+   * @return the command, for chaining.
+   */
+  description(t: string): Command {
+    this.descriptionText = t;
+    return this;
+  }
+
+  /**
+   * Sets any options for the command.
+   *
+   * @example
+   *   command.option("-d, --debug", "turn on debugging", false)
+   *
+   * @param args the commander-style option definition.
+   * @return the command, for chaining.
+   */
+  option(...args: any[]): Command {
+    this.options.push(args);
+    return this;
+  }
+
+  /**
+   * Attaches a function to run before the command's action function.
+   * @param fn the function to run.
+   * @param args arguments, as an array, for the function.
+   * @return the command, for chaining.
+   */
+  before(fn: ActionFunction, ...args: any[]): Command {
+    this.befores.push({ fn: fn, args: args });
+    return this;
+  }
+
+  /**
+   * Sets the help text for the command.
+   *
+   * This text is displayed when:
+   *   - the `--help` flag is passed to the command, or
+   *   - the `help <command>` command is used.
+   *
+   * @param t the human-readable help text.
+   * @return the command, for chaining.
+   */
+  help(t: string): Command {
+    this.helpText = t;
+    return this;
+  }
+
+  /**
+   * Sets the function to be run for the command.
+   * @param fn the function to be run.
+   * @return the command, for chaining.
+   */
+  action(fn: ActionFunction): Command {
+    this.actionFn = fn;
+    return this;
+  }
+
+  /**
+   * Registers the command with the client. This is used to inisially set up
+   * all the commands and wraps their functionality with analytics and error
+   * handling.
+   * @param client the client object (from src/index.js).
+   */
+  register(client: any): void {
+    this.client = client;
+    const program: CommanderStatic = client.cli;
+    const cmd = program.command(this.cmd);
+    if (this.descriptionText) {
+      cmd.description(this.descriptionText);
+    }
+    this.options.forEach((args) => {
+      const flags = args.shift();
+      cmd.option(flags, ...args);
+    });
+
+    if (this.helpText) {
+      cmd.on("--help", () => {
+        console.log(this.helpText);
+      });
+    }
+
+    cmd.action((...args: any[]) => {
+      const runner = this.runner();
+      const start = new Date().getTime();
+      const options = last(args);
+      const argCount = cmd._args.length;
+      if (args.length - 1 > argCount) {
+        client.errorOut(
+          new FirebaseError(
+            `Too many arguments. Run ${bold("firebase help " + cmd._name)} for usage instructions`,
+            { exit: 1 }
+          )
+        );
+        return;
+      }
```

Comment:
This block of code I very much don't like because it's depending on an internal property of `commander`. Would it be reasonable to remove this block? I could dig into what commander does if you give it too many arguments if that's interesting...

Reply:
I think that was put there for a reason, but I certainly don't remember what it was anymore. I'm curious what happens if you remove it and supply too many arguments. My guess is that it doesn't error and the extra args are just dropped, but given the way args are translated into the arguments for Command, it could cause damaging ripple effects.

Reply:
Yeah I also don't like this block ... plus if you look at the `Commander` release notes I linked above there are some mentioned changes to arg counting so these internals may not even be valid anymore.

Reply:
I'm working on this...

Reply:
Turns out the internals are still valid, but it's still not great...

Reply:
I MADE THIS BETTER! See the code for the solution and explanation.

Reply:
Whaa? I broke this. The emulators setup command broke. I'll look... thanks functional tests.

Reply:
Okay. The emulator setup functions were bound functions (which don't need options), which violates the rule and the requirement to actually accept options (which allows this new logic to work). I've updated them and everything should be happy now.